llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8702,61 +8738,264 @@ SDValue SystemZTargetLowering::combineSETCC(
-  // If the CCReg instruction is itself a ICMP testing the condition
+  // If the CCReg instruction is itself a ICMP / TM  testing the condition
   // code set by some other instruction, see whether we can directly
   // use that condition code.
-
-  // Verify that we have an ICMP against some constant.
-  if (CCValid != SystemZ::CCMASK_ICMP)
-    return false;
-  auto *ICmp = CCReg.getNode();
-  if (ICmp->getOpcode() != SystemZISD::ICMP)
-    return false;
-  auto *CompareLHS = ICmp->getOperand(0).getNode();
-  auto *CompareRHS = dyn_cast<ConstantSDNode>(ICmp->getOperand(1));
-  if (!CompareRHS)
+  auto *CCNode = CCReg.getNode();
+  if (!CCNode)
     return false;
-
-  // Optimize the case where CompareLHS is a SELECT_CCMASK.
-  if (CompareLHS->getOpcode() == SystemZISD::SELECT_CCMASK) {
-    // Verify that we have an appropriate mask for a EQ or NE comparison.
-    bool Invert = false;
-    if (CCMask == SystemZ::CCMASK_CMP_NE)
-      Invert = !Invert;
-    else if (CCMask != SystemZ::CCMASK_CMP_EQ)
+  if (CCNode->getOpcode() == SystemZISD::TM) {
+    if (CCValid != SystemZ::CCMASK_TM)
       return false;
-
-    // Verify that the ICMP compares against one of select values.
-    auto *TrueVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(0));
-    if (!TrueVal)
+    if ((CCMask != SystemZ::CCMASK_TM_SOME_1) &&
+        (CCMask != SystemZ::CCMASK_TM_ALL_0))
       return false;
-    auto *FalseVal = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(1));
-    if (!FalseVal)
+    SDValue Op0 = CCNode->getOperand(0);
+    SDValue Op1 = CCNode->getOperand(1);
+    auto [Op0CC, Op0CCValid] = findCCUse(Op0);
+    if (Op0CC == SDValue())
       return false;
-    if (CompareRHS->getAPIntValue() == FalseVal->getAPIntValue())
-      Invert = !Invert;
-    else if (CompareRHS->getAPIntValue() != TrueVal->getAPIntValue())
+    int TMConstVal;
+    auto *N = dyn_cast<ConstantSDNode>(Op1);
+    if (N) {
+      TMConstVal = N->getZExtValue();
+      if ((TMConstVal != (1 << SystemZ::IPM_CC)) && (TMConstVal != 1))
+        return false;
+    }
+    auto &&Op0CCVals = simplifyAssumingCCVal(Op0, Op0CC);
+    auto [Op1CC, Op1CCValid] = findCCUse(Op1);
+    auto Op1CCVals =
+        simplifyAssumingCCVal(Op1, Op1CC == SDValue() ? Op0CC : Op1CC);
+    if (Op0CCVals.empty() || Op1CCVals.empty())
       return false;
-
-    // Compute the effective CC mask for the new branch or select.
-    auto *NewCCValid = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(2));
-    auto *NewCCMask = dyn_cast<ConstantSDNode>(CompareLHS->getOperand(3));
-    if (!NewCCValid || !NewCCMask)
+    std::transform(Op0CCVals.begin(), Op0CCVals.end(), Op1CCVals.begin(),
+                   Op0CCVals.begin(), [&TMConstVal](auto &CC0, auto &CC1) {
+                     return (TMConstVal == (1 << SystemZ::IPM_CC))
+                                ? ((CC0 & CC1) >> SystemZ::IPM_CC)
+                                : (CC0 & CC1);
```

Comment:
No, this should implement the full TM logic as defined in the PoP:

- CC0: Selected bits all zeros; or mask bits all zeros
- CC1: Selected bits mixed zeros and ones, and leftmost is zero
- CC2: Selected bits mixed zeros and ones, and leftmost is one
- CC3: Selected bits all ones

Since we know both operands as constants, we can just implement this logic as-is.